sdk-lib/src/main/java/cash/z/ecc/android/sdk/block/processor/CompactBlockProcessor.kt

```diff
@@ -1946,7 +1951,10 @@ class CompactBlockProcessor internal constructor(
                 }
                 when (val response = downloader.fetchTransaction(rawTransactionId)) {
                     is Response.Success -> {
-                        transactionDataResult = response.result.data
+                        val currentMinedHeight =
+                            runCatching { response.result.height.toBlockHeight(network) }
```

Comment:
This possibly won't work as intended. The related gRPC layer says about the height parameter: `height that the transaction was mined (or -1)`. So, the exception will be thrown even before this check in parsing the fetch response on line 1953, where the response data is parsed to `RawTransactionUnsafe` in which we have `BlockHeightUnsafe` that has this check included:

```
data class BlockHeightUnsafe(val value: Long) : Comparable<BlockHeightUnsafe> {
    init {
        require(UINT_RANGE.contains(value)) { "Height $value is outside of allowed range $UINT_RANGE" }
    }

    override fun compareTo(other: BlockHeightUnsafe): Int = value.compareTo(other.value)

    companion object {
        private val UINT_RANGE = 0.toLong()..UInt.MAX_VALUE.toLong()
    }
}
```
We'll check the possible -1 for height, but in the wrapping retry logic. We could explicitly check the height parameter for `<= 0` and fill in the height as null in RawTransactionUnsafe instead. What do you think?

Reply:
Ah, I didn't realise that `BlockHeightUnsafe` had the same check. That means `BlockHeightUnsafe` is the incorrect type to use here, as there are three values that can be returned:
- 0 if the transaction is in the mempool.
- -1 (which is `0xffffffffffffffff` in the protobuf as it uses `uint64`) if the transaction is mined in a block that is not in the main chain.
- A non-negative height if the block is mined in the main chain.

So we'll need a new type that can specifically handle these semantics.

> The related gRPC layer says about the height parameter: `height that the transaction was mined (or -1)`.

This documentation was completely incorrect, as @nuttycom and I realised last week. The semantics are correctly documented, and bugs fixed, in https://github.com/zcash/lightwalletd/pull/495,